### DISCUSSION TITLE
how to save and restore opened tree

### DISCUSSION THREAD

**Discussion by mike-boost** [piroor/treestyletab#3053] [Q&A]
## Short description
I want to save all tabs and tree structure to bookmarks or file, and import from them.
I didn't find a clear way to do this.
I installed "Bookmark Tree for Tree Style Tab by Lej77", but didn't find the guide.

- **Bookmark Tabs...**  and **Bookmark Selected Tree...** , no response.
- **Bookmark Tree Data** only save tabs in bookmark without structure

![image](https://user-images.githubusercontent.com/69611731/145013015-003e7580-dc71-4cf5-b635-572f6f197eab.png)

## Expected result
Have a guide to explain how to save and restore opened tree.

## Environment

 * Platform (OS): Mac bigsur
 * Version of Firefox: 94.0.2 (64-bit)
 * Version (or revision) of Tree Style Tab: 3.8.16 (2021.11.11):

**Comment by irvinm** [comment]
It sounds like you don't have the bookmark permission set for TST.  Do you have this enabled in TST?  (I don't and I get a notification when I try to bookmark tabs from TST to enable the permission first)  I'm not sure how this differs on the mac.

![image](https://user-images.githubusercontent.com/979729/147357981-ad8fc5e2-8f21-4118-81a7-78853b3b8e99.png)

**Reply by mike-boost** [reply]
> It sounds like you don't have the bookmark permission set for TST. Do you have this enabled in TST? (I don't and I get a notification when I try to bookmark tabs from TST to enable the permission first) I'm not sure how this differs on the mac.
> 
> ![image](https://user-images.githubusercontent.com/979729/147357981-ad8fc5e2-8f21-4118-81a7-78853b3b8e99.png)

Thanks.
I enable thoese and now it can save book marks, but still not the structure.  It only rename the tabs to pretend they have structure.

<img width="730" alt="image" src="https://user-images.githubusercontent.com/69611731/165279527-de44e94f-0804-42a6-94b3-e73862aac3d8.png">

when save tree:
<img width="378" alt="image" src="https://user-images.githubusercontent.com/69611731/165280601-003e2873-e8f4-477e-b1c9-c08404069edd.png">


and open by firefox bookmark open all:
<img width="252" alt="image" src="https://user-images.githubusercontent.com/69611731/165280669-df8708ae-ffd0-4945-a62d-89b42f69510a.png">

maybe I open in a wrong way 😏

**Reply by SXZ1** [reply]
@mike-boost open bookmarks, right-click on a folder containing TST tree structure and click on Open all as tree or something like that, this menu item should have a TST icon.

**Reply by mike-boost** [reply]
> @mike-boost open bookmarks, right-click on a folder containing TST tree structure and click on Open all as tree or something like that, this menu item should have a TST icon.

Thanks.
It only clickable in the opened Bookmarks window, not available in dropdown bookmark menu!

**Reply by GlenCoakley** [reply]
Whether invoked from the bookmarks manager or a bookmarks pop-up, this will create a Firefox tab group then open all bookmarks from the saved tree as direct children in that group. It does not preserve a tree with multiple levels/depths. 
I believe the expectation of the OP (and certainly my expectation) is that it would preserve the full, complex structure of selected trees of tabs.

If I have tabs as represented here (pretending this is a vertical view of tabs in the Tree Style Tab sidebar.
```
  tab with title "ABC", which is the top-level, parent tab
    tab with title "DEF" which is second level down, "child #1", of parent
      tab with title "GHI" which is third level down, child of "child #1" and "child #1.1" of parent
  tab with title "JKL" which is second level down, "child #2" of parent
     tab with title "MNO" which is third level down, child of "child #2" and "child #2.1" of parent
```

My expectation is to see such a set of trees within trees to be duplicated when the tabs are reopened. Instead what is created is the following.
 ```
  tab group
    tab with title "ABC" which is second level down, child #1 in group
    tab with title "DEF" which is second level down, child #2 in group
    tab with title "GHI" which is second level down, child #3 in group
    tab with title "JKL" which is second level down, child #4 in group
    tab with title "MNO" which is second level down, child #5 in group
```

Currently, this extension will do what is desired: https://github.com/Lej77/bookmark-tree-for-tree-style-tab
